### ISSUE TITLE
Unlabeled buttons

### ISSUE THREAD

**Issue by joshms123** [home-assistant/home-assistant-polymer#3267]
**Home Assistant release with the issue:**
0.94.1

**Last working Home Assistant release (if known):**
None

**UI (States or Lovelace UI?):**
Lovelace

**Browser and Operating System:**
Chrome, Windows 10

**Description of problem:**

As a visually impaired user, I have found several accessibility issues over the year or so while using home assistant. The biggest issue currently (of mine at least) is the number of unlabeled buttons in the Lovelace UI. Currently when navigating with my screen reader, I hear "unlabeled 0", "unlabeled 1", etc. With a different screen reader (NVDA), I get even less feedback. 
List of unlabeled buttons I know of currently:
- The button to show the menu bar and developer tools.
- Notifications button.
- Button to activate the microphone.
- Button that shows a menu with unused entities, configure UI and help.
- Button after that is unknown (nothing noticeable seems to happen).
- Almost all buttons to control devices are unlabeled  apart from execute for scripts, learn more, ARM HOME/AWAY and lock/unlock. I may have missed a few here.
This is the current list of buttons I have just gone over; if I find more, I will either update or comment.

**Javascript errors shown in the web inspector (if applicable):**
N/A
```

```

**Additional information:**
N/A

**Comment by balloob** [comment]
I have opened a pull request to address a couple of the mentioned missing labels. The pull request is #3310.

**Comment by balloob** [comment]
For people looking into helping with this issue, there is an accessibility dev tool to help find bad buttons: https://www.deque.com/axe/.

**Comment by steverep** [comment]
I am a visually impaired user as well and noticed media device buttons, as well as a few others, got labeled in a recent release.  Thanks for that.

My cover device cards are not labeled yet though.

I'll point out more as I come across them.

**Comment by balloob** [comment]
Opened pull request #3421 to add labels to the cover and cover tilt controls.

**Comment by joshms123** [comment]
Hi there!

Great work on the accessibility fixes so far, I really appreciate it!

The following buttons are unlabeled:
- On the hass.io tab, when viewing an add-on the back button is unlabeled.
- When viewing more information in the states view, the close/dismiss button is unlabeled.

The following button from my original post is still unlabeled:
- Button to activate the menu that shows unused entities, configure UI and help.

**Comment by balloob** [comment]
@joshms123 addressed in #3422

**Comment by steverep** [comment]
In the modal dialog when editing an entities card, there are 2 unlabeled buttons associated with each entity in the list.  One of them deletes the entity and I'm not sure what the other one does.

**Comment by balloob** [comment]
@steverep addressed in #3430

**Comment by steverep** [comment]
Two more:
- [ ] Close button on any modal dialog (e.g. device more info, adding integration, editing view, etc.)
- [ ] When configuring Lovelace UI, button after each card that shows a menu to either move or delete the card

**Comment by bramkragten** [comment]
Do you have a more specific example of the close button?
I don't see a close button in the adding integration and the more info dialog does have a label. 
The view editing dialog only has a cancel button, but that has text in it so should be ok?

**Comment by steverep** [comment]
> Do you have a more specific example of the close button?

Sorry, I was a bit lax on that and looks like several were fixed.  Here's a few that I just found:

- [ ] Close button on notifications dialog
- [ ] When opening an entity's more info dialog from a view, there is an unlabeled button that opens another dialog to edit the entity name and ID
- [ ] The resulting edit name/ID dialog has a close button that is also unlabeled.  Weirdly, it is labeled if you go through the entity registry instead.

**Comment by balloob** [comment]
@steverep wanted to let you know that @bramkragten is working for Nabu Casa and will focus on the Home Assistant frontend. Improving accessibility is one of this tasks so please keep the issues coming.

**Comment by steverep** [comment]
A few more to add to the list from my last comment:
- [ ] Button to show menu with Configure UI and Help
- [ ] Button to exit UI configuration mode
- [ ] Button to edit Lovelace UI title

**Comment by iantrich** [comment]
@balloob @bramkragten is there a way to integrate axe into our test builds to alert us before merging new unlabeled buttons?

**Comment by steverep** [comment]
More...

- [ ] On Configuration > Integrations page, buttons/links to devices under each integration are all unlabeled

**Comment by steverep** [comment]
- [ ] The button to add an integration is now labeled "hass:plus".  I'm pretty sure this used to to have a useful label, so some extra quality control on new commits would help.

**Comment by springstan** [comment]
Can we somehow translate the aria-labels in js files? 
So far I had no luck and I was wondering if there was any way to do so.

**Comment by iantrich** [comment]
@springstan only if it extends the LocalizeMixin, I believe

**Comment by springstan** [comment]
@iantrich I thought so too, but as soon as I use localize with the aria-label the attribute disappears.

**Comment by bramkragten** [comment]
@springstan you should use the `$` after the attribute name in Polymer so it will be an attribute, otherwise Polymer makes it a property.

```javascript
          aria-label$="[[localize('ui.translation.key')]]"
```

**Comment by nickallan1** [comment]
Hi
just started using homeassistant and I'm also a screen reader user.
Some unlabelled controls I've found
developer tools
there is a left arrow before the tabs and a right arrow after the tabs. these are visually disabled but are still active for a screen reader as the disabled attribute hasn't been set, they also need to have text labels.

Interesting, until you activate a tab control, the screen reader doesn't see its label.
e.g
I've got the MQTT tab, but until I activate that tab, the "MQTT" text is not visible to my screen reader.
Once you've activated a tab once, the text label is then visible.


service tab
when a value is in the service edit dropdown, the clear button that appears needs a label.
there's also a second button that shows up there as well, not sure what it does.

screen reader being used for testing is Jaws 2020 from www.freedomscientific.com

**Comment by nickallan1** [comment]
Another issue I've found is in the sidebar.
when you look at this with a screen reader there are 2 controls, the first is a button that has a label "Sidebar toggle" and next is a listbox. If you use the up and down arrows in the listbox before pressing the sidebar toggle, the screen reader announces the url for each item in the list instead of the item's text. e.g my screen reader announces the following as you down arrow
 Link http://hassio.local:8123/logbook
Link http://hassio.local:8123/history
Link http://hassio.local:8123/core_configurator

If I press the sidebar toggle button and then use the down arrow in the listbox, the screen reader says
Logbook Link 
History Link 
Configurator Link
The issue is that the Sidebar toggle button doesn't indicate its state, so you can't tell if your in the mode where the listbox will speak correctly or not until you've started arrowing through the list items.
suggestion
1. just make the listbox work correctly all the time without needing to press the sidebar toggle button.

or
 2. use the aria-expanded attribute on the sidebar toggle button to indicate if the sidebar is visible or not.
if this is the preferred solution, you probably want to actually hide the listbox element from screen readers if the toggle button is in its collapsed state.
https://www.w3.org/WAI/GL/wiki/Using_aria-expanded_to_indicate_the_state_of_a_collapsible_element

**Comment by bramkragten** [comment]
Thanks for reporting @nickallan1! I'll look into this shortly.

**Comment by scottyphillips** [comment]
Hi Devs, @nickallan1 also mentioned the Thermostat card needs to be looked into as well as it relates to a custom component of mine. (see https://github.com/scottyphillips/mitsubishi_echonet/issues/9) 

I opened up issue https://github.com/home-assistant/home-assistant/issues/30495 but in hindsight should have checked here first.

**Comment by nickallan1** [comment]
thanks @bramkragten
in relation to the thermostat controls in lovelace. as noted above the buttons aren't labelled. I assume there's supposed to be a slider to allow changing of temperature, its also not accessible.

**Comment by bramkragten** [comment]
For the thermostat slider, @thomasloven can probably help with that?

**Comment by steverep** [comment]
@bramkragten the last 2 check boxes above don't look to be fixed in the latest release but you checked them off.  And while I'm at it...

- [ ] Button to add an area is incorrectly labeled "hass plus" as well

**Comment by steverep** [comment]
As far as I can tell, all buttons have now been replaced with labels in the form "hass:icon-name" in the latest release.

I'm losing confidence in Home Assistant taking accessibility seriously.  Is there a goal to close this issue and ensure new releases have appropriately labeled buttons or not?  Providing alternative text for graphical elements is literally criterion 1.1.1 in WCAG.  There are a lot more accessibility issues to report.

**Comment by bramkragten** [comment]
That is something one of our upstream dependencies does, I will fix that.
We made some switches on upstream elements we use and this one does things differently than the one we used before.

Thanks for letting us know this so we can fix it!

**Comment by bartbunting** [comment]
Hi, another screenreader user here.
Having read the comments here I'm still encountering many buttons reporting as "Unlabeled <number>".
frequently it is what I can only assume is the "Back" button as it appears to perform that function.
Also on the Configuration > Integrations page the button to add a new integration is still read as "Unlabeled 0".
I'd be happy to try and fix some of these but have no idea where I should be looking as I don't understand how they are being created.  Guessing by some library as grep doesn't find them.
Would very much appreciate this being addressed.

**Comment by spacegaier** [comment]
With my PR https://github.com/home-assistant/frontend/pull/7587 all fab buttons (used e.g. to setup new integrations, create new automations or save changed scripts) should now provide an `aria-label`.

**Comment by nickallan1** [comment]
This isn't working for me. Looking at the DOM, It would seem to be because the actual button element has aria-label="". the aria-label that has the correct button text is a parent of the button. At least Jaws is looking at the aria-label="" on the button that gets focus so it calls it unlabelled0. This is the same type of issue you see in the lovelace controls as well in most cases.
Note, if you turn jaws's virtual buffer off and tab through the UI, it does seem to pick up the aria-label that actually has a text value, but in the normal virtual buffer interface, most things seem to still be unlabelled.
Parcial example of the setup integration button below.

```
<mwc-fab slot="fab" aria-label="Set up a new integration" title="Set up a new integration">
<button class="mdc-fab  " aria-label="">
        
        <mwc-ripple class="ripple"></mwc-ripple>
        
        <span class="icon-slot-container">
          <slot name="icon">
            <!----><!---->
          </slot>
        </span>
        <!----><!---->
        <!----><!---->
      </button>
        </mwc-fab>```

**Comment by spacegaier** [comment]
@nickallan1 Are you referring to my PRs or in general to this topic? With my changes from PR #7587, I get the following DOM. The `<button>` does have the `aria-label` filled properly.

![grafik](https://user-images.githubusercontent.com/114137/98039432-fe588b80-1e1e-11eb-88f1-559cd7c5d5ec.png)

The DOM you posted is the old, before situation.

**Comment by nickallan1** [comment]
@spacegaier 
Is this supposed to be in the current production release? I'm running the following versions
Home Assistant 0.117.2
Frontend version: 20201021.4 - latest
host_os HassOS 4.15
supervisor 2020.10.1

**Comment by spacegaier** [comment]
@nickallan1 No, not yet. My changes are a pull request (= coding proposal) that I submitted earlier today and still needs to be reviewed and merged into the official coding version. So it will hopefully be part of the next beta versions.

**Comment by bartbunting** [comment]
@spacegaier I just tested and your change makes the button to add a new integration speak correctly.
I am still seeing an "Unlabled 0" for what appears to be the back arrow.
This appears both on the Integrations page itself and also at the top of add a new integration page.
This is a great improvement though.  Haven't been able to test every instance you updated but will try and test further.

**Comment by spacegaier** [comment]
Just to make sure: You are aware that the PR for the fabs (the "new integration" for example) and the one for the backend buttons are two separate ones? Did you test both?

**Comment by bartbunting** [comment]
Sorry my bad, I didn't notice that.
I'm still getting my head around how to test and I think my way is likely not the best as yet.
Having checked out both branches of your repo I can report that yes they both do different things.
fab-aria-labels fixes the Setup new integration" button and friends.
The aria-prev-next branch labels the "back" button on the config > integrations screen (and presumably in other places it's used).
 
However there is still a button with "Unlabeled 0" at the top of the model for adding a new integration.  Guessing it's meant to be close or similar?

This is great progress thanks!

I am happy to test and report more of these issues, just need to learn a smoother workflow for testing pull requests and make better friends with the chrome devtools using jaws.

**Comment by spacegaier** [comment]
Yes, the close button of that particular dialog is not included in those PRs, I will create another one for all `<ha-icon-button>`. 102 instances to check and potentially adjust in the source.

**Comment by Nardol** [comment]
Hello,

I noticed another group of buttons which are not labelled: all button for Google cast devices.
Also when assigning a device to a room, Orca announce a button with no label, which I cannot test because the focus return to the edit box to specify the room.

I use Orca on Debian Buster.

**Comment by joshms123** [comment]
Hello there,
There are still quite a few unlabeled buttons in the current home assistant release; including lovelace cards, configuration, supervisor, ETC.
I would like to assist in resolving these, firstly, is there a way to easily edit the HTML where these buttons are?
Secondly, is there possibly an automated testing tool that could look for these issues to assist developers in resolving these issues in their code, before its published to release?
Thanks,
Josh.

**Comment by spacegaier** [comment]
I am working on it, but it is a big task, since we have a lot of icon buttons throughout HA: https://github.com/home-assistant/frontend/pull/9230

**Comment by steverep** [comment]
Again, I'm going to ask if there is ever going to be a serious (or at least half-heart) attempt to ensure accessibility of Home Assistant? This issue is well over 2 years old and again I'll mention it's criterion 1.1.1 on accessibility guidelines.

I'd love to just list problem buttons I find as I have done earlier in this thread a number of times, but some of the buttons already fixed were then unlabeled again in a future release, so I don't want to waste my time.

Surely the developers could write a simple unit test to check all pull requests for unlabeled buttons before allowing a merge.

**Comment by joshms123** [comment]
Hi,
seems like #9230 was merged. I know there are still a few problems, but overall I have noticed quite an improvement.

**Comment by steverep** [comment]
Is that PR in the next release?

**Comment by spacegaier** [comment]
> Is that PR in the next release?

That was already part of the current 2021.11 release.

> I'd love to just list problem buttons I find as I have done earlier in this thread a number of times, but some of the buttons already fixed were then unlabeled again in a future release, so I don't want to waste my time.
> 

We can only fix what we are aware of. Overall my above mentioned PR generalized the handling of all icon-buttons, so a lot more of them should now have an ARIA label assigned. But for sure there are still a few, that need to be onboarded, so reporting them makes sense.

> Again, I'm going to ask if there is ever going to be a serious (or at least half-heart) attempt to ensure accessibility of Home Assistant? 

We are always trying to ensure accessibility, but of course that competes as with all requests for the limited resources of the project. but just yesterday there was another UX meeting reg. accessibility, so those discussions are happening (see the "#devs_ux" Discord channel).

**Comment by steverep** [comment]
> We can only fix what we are aware of. Overall my above mentioned PR generalized the handling of all icon-buttons, so a lot more of them should now have an ARIA label assigned. But for sure there are still a few, that need to be onboarded, so reporting them makes sense.

In that case, in the spirit of specifics and not just complaining:
- [ ] Buttons to open/stop/close a cover
- [ ] Buttons on automations configuration page in the last column of the table (3 buttons for each automation)

There's more... I just can't think of them off the top of my head.  I'll find them.

**Comment by spacegaier** [comment]
I created the PR for the automation page (found two more icon buttons on that page that had issues that will be fixed by the PR).

**Comment by spacegaier** [comment]
Second raised point (well actually the first in your two point list 😄), will be solved by PR #10859.

**Comment by steverep** [comment]
As promised:
- [ ] On the ZWave-JS integration configuration page, buttons to select log level and download log

**Comment by Nardol** [comment]
Thermostat card buttons are also not labelled.
```yaml
- type: thermostat
  entity: climate.thermostat_salon
```

**Comment by spacegaier** [comment]
PR #10875 fixes the climate modes in the thermostate card.

PR #10876 fixes the log download button for ZWave JS. I however have no such setup, so this is untested. 

Reg. the log levels: Could you please add a screenshot so that I can see which element this is regarding, because just looking at the code I did not see anything there.

**Comment by steverep** [comment]
One more:
- [ ] On The system configuration page, there's a "more options" button after the host shutdown in the tab sequence.  It shows hardware, import from USB, and move datadisk as options.

> Reg. the log levels: Could you please add a screenshot so that I can see which element this is regarding, because just looking at the code I did not see anything there.

Upon closer inspection, it's not actually a button but instead a drop-down menu to select the log level.  There's a similar one on the system configuration page as well, and a lot more now that I think about it.  The problem is that their ARIA is very screwy (e.g. they are being exposed mixed as buttons and menus, tabbing works but browsing doesn't, etc.).  I'll tkae a closer look at the HTML and open a separate ticket to work on the ARIA pattern.

**Comment by Nardol** [comment]
Thanks!

I also noticed the buttons are not labelled on the `media_player` card.
```yaml
- type: media-control
  entity: media_player.enceinte_bureau
```

**Comment by spacegaier** [comment]
@steverep PR created. Separate issue for the dropdowns is the correct way 👍 

@Nardol PR created as well.

**Comment by Nardol** [comment]
Thanks for your work!
In the visualization screen of a climate entity, the two button to set up and down the temperature from 0.5° are not labelled.
I know what they do because I pressed these :slightly_smiling_face: I made a snapshot to be sure I explained which screen right.
![screen_thermostat](https://user-images.githubusercontent.com/2864821/146042152-645fa597-a0b1-4819-b554-a980227bd8a5.png)

**Comment by spacegaier** [comment]
@Nardol: Good catch. PR #10921 will fix that.

**Comment by Nardol** [comment]
I also noticed buttons on and of for switches with assumed_state: true are also not labelled.
Next time I'll try to have more than one...

**Comment by spacegaier** [comment]
> I also noticed buttons on and of for switches with assumed_state: true are also not labelled. Next time I'll try to have more than one...

Can you provide some more details what you mean, so I can try to reproduce.

**Comment by Nardol** [comment]
For example Broadlink RM Mini allow to create switch to execute remote commands.
In this case, instead of a classic switch we have two buttons: one for switching it on and the otther to switch it off.
And these buttons are not labelled.
These switch have the property assumed_state set to on.

**Comment by spacegaier** [comment]
> For example Broadlink RM Mini allow to create switch to execute remote commands. In this case, instead of a classic switch we have two buttons: one for switching it on and the otther to switch it off. And these buttons are not labelled. These switch have the property assumed_state set to on.

And those are icon buttons so buttons that only show an icon, without any text? Can you quickly add a screenshot to illustrate and ensure we are talking about the same thing 😄.

**Comment by Nardol** [comment]
Hm. My mistake, I should have tested better. It was something I did not use for some weeks and remembered.
Looks OK now, sorry.

**Comment by steverep** [comment]
In debugging #8178 and #10617, I noticed another problem with nearly all buttons I've seen across the UI: the SVG icon within the button has no accessible name, which fails [WCAG criterion 1.1.1](https://www.w3.org/TR/WCAG21/#non-text-content).

In most cases I'm seeing the `ha-icon-button` (wrapped around `mwc-icon-button`) or `ha-fab` components used to insert `<button>` elements with SVG children.  Since the `<button>` element is getting the `aria-label` for `ha-icon-button`, the child icon should be removed from the accessibility tree.  Giving it a role also doesn't hurt since SVG has no implicit role.  So please add the 2 attributes as:

`<svg role=img" aria-hidden="true" ...`

In the case of the `ha-fab` component, the same applies for the SVG, but also the `aria-label` is completely unnecessary since the accessible name will be taken from the text.  There's no harm if they match, but if one were to change without the other this might run afoul of [WCAG criterion 2.5.3](https://www.w3.org/TR/WCAG21/#label-in-name).

**Comment by Nardol** [comment]
After some test :slightly_smiling_face: I found two unlabelled check boxes:
In configuration->Integrations when clicking system configuration for an integration "Enable newly added entities" and "Enable polling for updates." are not labelled right for screen readers, tested on Firefox.
If I press down arrow I have the text but not when pressing tab/Shift+Tab to focus these settings, instead I have "[object Object]".
The label are not directly set for the check boxes.

I would like to contribute and tried to understand why the labels were not right but I did not figure how to do so, where or how could I find some way to help?

**Comment by spacegaier** [comment]
@steverep PR https://github.com/home-assistant/frontend/pull/11389 created for the SVG topic. 

Reg. `<ha-fab>`: We are not specifically setting an `aria-label` here, but just the regular `label` of the component that then automatically gets also set apparently as the `aria-label`. So they should always match.

**Comment by steverep** [comment]
Thanks @spacegaier.  I recently setup a a complete dev environment for myself so I can fix a lot of that small stuff myself now.  I could really use your help with #8178 if you have time to lend.

**Comment by spacegaier** [comment]
> Thanks @spacegaier. I recently setup a a complete dev environment for myself so I can fix a lot of that small stuff myself now. I could really use your help with #8178 if you have time to lend.

I don't own a single Apple device, so I am afraid I cannot really help with #8178

**Comment by steverep** [comment]
Finding a few more while fixing other accessibility issues...
- [ ] Close button for hardware dialog (looks like a bad variable reference)
- [ ] More options popup menu button on Backups tab
- [ ] More options popup menu on System tab under Host
